types/mapper.go

```diff
@@ -0,0 +1,232 @@
+package types
+
+import (
+	"reflect"
+	"fmt"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	crypto "github.com/tendermint/go-crypto"
+	amino "github.com/tendermint/go-amino"
+)
+
+// Manages utxo's in existence
+// Uses go-amino encoding/decoding library
+// Does not need to be changed to RLP
+type UTXOMapper struct {
+	
+	// The key used to access the store from the Context.
+	key sdk.StoreKey
+
+	// The prototypical UTXO concrete type
+	proto UTXOHolder
+
+	// The Amino codec for binary encoding/decoding of utxo's 
+	cdc *amino.Codec 
+}
+
+// NewUTXOMapper returns a new UtxoMapper that
+// uses go-Amino to (binary) encode and decode concrete UTXO
+func NewUTXOMapper(key sdk.StoreKey, proto UTXOHolder) UTXOMapper {
+	cdc := amino.NewCodec()
+	Register(cdc)
+	return UTXOMapper {
+		key:   key,
+		proto: proto,
+		cdc:   cdc, 
+	}
+
+}
+
+// Create and return a sealed utxo mapper. Not sure if necessary
+func NewUTXOMapperSealed(key sdk.StoreKey, proto UTXOHolder) sealedUTXOMapper {
+	cdc := amino.NewCodec()
+	um := UTXOMapper {
+		key:   key,	
+		proto: proto,
+		cdc:   cdc,
+	}
+	// Register for amino encoding/decoding
+	Register(cdc)
+
+	// make accountMapper's AminoCodec() inaccessible
+	return um.Seal()
+}
+
+// Register all crypto interfaces and concrete types necessary
+func Register(cdc *amino.Codec) {
+	cdc.RegisterConcrete(crypto.PubKey{}, "go-crypto/PubKey", nil)
+	cdc.RegisterConcrete(crypto.PrivKey{}, "go-crypto/PrivKey", nil)
+	cdc.RegisterConcrete(crypto.Signature{}, "go-crypto/Signature", nil)
+	cdc.RegisterConcrete(sdk.StdSignature{}, "sdk/StdSignature", nil)
+	cdc.RegisterInterface((*crypto.PubKeyInner)(nil), nil)
+	cdc.RegisterConcrete(crypto.PubKeyEd25519{}, "go-crypto/PubKeyEd25519", nil)
+	cdc.RegisterConcrete(crypto.SignatureEd25519{}, "go-crypto/SignatureEd25519", nil)
+	cdc.RegisterInterface((*crypto.SignatureInner)(nil), nil)
+	cdc.RegisterInterface((*UTXOHolder)(nil), nil)
+	cdc.RegisterConcrete(BaseUTXOHolder{}, "types/BaseUTXOHolder", nil)
+}
+
+// Returns the go-Amino codec.
+func (um UTXOMapper) AminoCodec() *amino.Codec {
+	return um.cdc
+}
+
+// Returns a "sealed" utxoMapper
+// the codec is not accessible from a sealedUTXOMapper
+func (um UTXOMapper) Seal() sealedUTXOMapper {
+	return sealedUTXOMapper{um}
+}
+
+func (um UTXOMapper) GetUTXO(ctx sdk.Context, addr crypto.Address, denom uint64) UTXO {
+	store := ctx.KVStore(um.key) // Get the utxo store
+	bz := store.Get(addr) // Gets the encoded bytes at the address addr
+	// Checks to see if there is a utxo at that address
+	if bz == nil {
+		return nil
+	}
+	utxoHolder := um.decodeUTXOHolder(bz) // Decode the go-amino encoded utxoHolder
+	utxo, _ := utxoHolder.GetUTXO(denom) // Get the utxo from utxoHolder
+	return utxo
+}
+
+func (um UTXOMapper) CreateUTXO(ctx sdk.Context, utxo UTXO) {
+	addr := utxo.GetAddress() // Get the address of the utxo
+	store := ctx.KVStore(um.key) // Get the utxo store 
```

Comment:
What's the underlying db structure used to store UTXOs? Is it just a key-value store with an Address mapped to a UTXOHolder?

Reply:
Yup. I originally had an address mapped to an UTXO, but the issue I found was an address could only have 1 utxo in that case. In plasma it should not be an issue to have outputs be sent to the same address since they are all uniquely identified by the blocknumber and tx index they are included in. So my solution to this was to create a structure which holds all the utxo's at a specific address. I think there may be a better way of handling this however. What are your thoughts?